windows/process.mpl

```diff
@@ -1,159 +1,143 @@
 
-  terminate: [
-    exitCode:;
-    result: exitCode processInfo.hProcess TerminateProcess;
-    result 0 "TerminateProcess" getErrorMessage
+  close: [
+    processInformation.hProcess CloseHandle 0 = ~ [processInformation.hThread CloseHandle 0 = ~] &&
   ];
 
-  getExitCode: [
-    out: {
-      exitCode: Nat32;
-      error:    String;
-    };
-
-    result:
-      out.exitCode storageAddress
-      processInfo.hProcess
-      GetExitCodeProcess;
-
-    result 0 "GetExitCodeProcess" getErrorMessage @out.!error
-
-    @out
-  ];
-
-  private getErrorMessage: [
-    code: failCode: functionName:;;;
-    code failCode = ~ [String] [
-      (functionName " failed, error code: " GetLastError) assembleString
-    ] if
-  ];
-
-  private DIE: [
-    (
-      processInfo.hProcess
-      processInfo.hThread
-      startupInfo.hStdInput
-      startupInfo.hStdOutput
-      startupInfo.hStdError
-    ) toIter [CloseHandle drop] eachStatic
-
-    1n32 terminate drop
-  ];
-
-  private INIT: [
-    STARTUPINFOW        !startupInfo
-    PROCESS_INFORMATION !processInfo
+  exitCode: [
+    out:;
+    errorTrait: @out processInformation.hProcess GetExitCodeProcess;
+    errorTrait 0 = ~
   ];
 }];
+
+# Start 'process' represented by 'command'
+# in:
+#   command
+# out:
+#   process
+#
+# Examples:
+#   1) "program" startProcess.closeInactive
+#
+#   2) "program programArgument" startProcess.closeInactive
+#
+#   3) program: "program" startProcess;
+#      program.operationSucceed [program.closeInactive] [
+#        ("Failed to start program: " program.operationError) printList
+#      ] if
+#
+#   4) program: "program" startProcess;
+#      program.tillActive
+#      exitCode: program.exitCode;
+#      program.close
+startProcess: [
+  command:;
+  {
+    process:            Process;
+    command:            command utf16;
+    operationErrorCode: Nat32;
+    toBeClosed:         Cond;
+
+    operationError: [operationErrorCode new];
+
+    operationSucceed: [operationError 0n32 =];
+
+    updateOperationError: [
+      determiner:;
+      determiner ~ [GetLastError !operationErrorCode] when
+    ];
+
+    doNotClose: [FALSE !toBeClosed];
+
+    tillActive: [process.tillActive updateOperationError];
+
+    closeInactive: [
+      tillActive
+      operationSucceed [close] when
+    ];
+
+    close: [
+      process.close updateOperationError
+      operationSucceed [doNotClose] when
+    ];
+
+    # NOTE: There is a corner case. If process did exit with status code 259, the function will report that the process is active
+    stillActive: [exitCode STILL_ACTIVE =];
+
+    exitCode: [
+      status: Nat32;
+      [@status process.exitCode] "[exitCode] failed" ensure
+      status
+    ];
+
+    INIT: [];
+
+    DIE: [
+      [toBeClosed ~] "Handles are not closed" ensure
+    ];
+
+    [
+      @process.init
+      @command @process.start updateOperationError
+      operationSucceed !toBeClosed
+    ] call
```

Comment:
Why is the code block needed? The code does not pollute the object.

Reply:
> why

Because we don't want to make mistakes later, when we'll change the code.